### PR TITLE
Implement JPA DataStore

### PR DESCRIPTION
**Pull request by wcekan**
Rebased from #651

DataStore with only javax.persistence dependency.
Can be used with any JEE container.
Supports both JTA and Non-JTA transactions.

### CODE REVIEW

**Comment by wcekan** [comment]
@belovaf @aklish Please review the rebase and merge.  

For the next step I would like to try to move the common code to shared common superclasses that will support any JPA library.  Hibernate 3 and 5 should simply subclass and add the required wrappers.

File: elide-datastore/elide-datastore-jpa/pom.xml [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>elide-datastore-jpa</artifactId>
+    <packaging>jar</packaging>
+    <name>Elide Data Store: JPA</name>
+    <description>Elide Data Store for JPA support</description>
+    <url>https://github.com/yahoo/elide</url>
+    <parent>
+        <groupId>com.yahoo.elide</groupId>
+        <artifactId>elide-datastore-parent-pom</artifactId>
+        <version>4.3.1-SNAPSHOT</version>
```
Comment (aklish) [inline elide-datastore/elide-datastore-jpa/pom.xml]: This needs to be updated to 4.3.3

File: elide-datastore/elide-datastore-jpa/pom.xml [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    <dependencies>
+        <!-- Elide -->
+        <dependency>
+            <groupId>com.yahoo.elide</groupId>
+            <artifactId>elide-core</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.yahoo.elide</groupId>
+            <artifactId>elide-datastore-hibernate</artifactId>
+            <version>4.3.1-SNAPSHOT</version>
```
Comment (aklish) [inline elide-datastore/elide-datastore-jpa/pom.xml]: Same comment.

File: elide-datastore/elide-datastore-jpa/pom.xml [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            <artifactId>javax.transaction-api</artifactId>
+            <version>1.2</version>
+            <scope>provided</scope>
+            <optional>true</optional>
+        </dependency>
+
+        <dependency>
+            <groupId>com.yahoo.elide</groupId>
+            <artifactId>elide-integration-tests</artifactId>
+            <version>4.3.1-SNAPSHOT</version>
```
Comment (aklish) [inline elide-datastore/elide-datastore-jpa/pom.xml]: Same

File: elide-datastore/elide-datastore-jpa/src/test/java/com/yahoo/elide/datastores/jpa/JpaDataStoreSupplier.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    private static final String JDBC = "jdbc:h2:mem:root;IGNORECASE=TRUE";
+    private static final String ROOT = "root";
+
+    @Override
+    public DataStore get() {
+        Map<String, Object> options = new HashMap<>();
+        ArrayList<Class<?>> bindClasses = new ArrayList<>();
+
+        try {
+            bindClasses.addAll(ClassScanner.getAnnotatedClasses(Parent.class.getPackage(), Entity.class));
```
Comment (aklish) [inline elide-datastore/elide-datastore-jpa/src/test/java/com/yahoo/elide/datastores/jpa/JpaDataStoreSupplier.java]: I think you'll need more than this now.  This only covers one package.  You will need to include elide-example-models as well.

**Comment by aklish** [comment]
We should switch Elide-Standalone to use this along with the H2 in memory DB.

File: elide-datastore/elide-datastore-jpa/pom.xml (hunk 1 above)
Comment (wcekan) [reply elide-datastore/elide-datastore-jpa/pom.xml]: Done in the merge.

**Pull request merged**